### ISSUE TITLE
Using node-libcurl with AWS Lambda

### ISSUE THREAD

**Issue by trangiiee** [JCMais/node-libcurl#105]
Hi there, 

I'm attempting to use this node-libcurl with AWS Lambda. Locally, my code runs great, but when I deploy it, I'm getting this error: 

```
module initialization error: Error
    at Error (native)
    at Object.Module._extensions..node (module.js:597:18)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/var/task/node_modules/node-libcurl/lib/Easy.js:40:29)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
```

I'm very new to Lambda, but I think I need to pre-compile the code. I'm unsure how to configure the library to use the Linux 64 bit binary. 

Is it possible you could assist me?

Thanks!

Angie

**Comment by JCMais** [comment]
Hi, I'm afraid that I cannot help you much, since I don't have previous experience with AWS Lambda.

But from what I could find, this is probably going to help ya: https://aws.amazon.com/pt/blogs/compute/nodejs-packages-in-lambda/

**Comment by trangiiee** [comment]
Sounds good. Thanks for the tip!

**Comment by bairathirahul** [comment]
I was also trying to deploy this package to AWS lambda and came across the same issue. Here are my steps to deal with that - 

**TL;DR** - Build the node_libcurl.node binding using a AWS Lambda environment (https://github.com/lambci/docker-lambda), and copy it into the `./node_modules/node-libcurl/lib/binding` folder before deployment. 

**What's the issue**
I am using Mac OS for the development. Therefore `npm install` installs the libcurl binding specific to Mac. For the deployment, serverless cli simply packages the local project (including local modules) and deploys it. Hence, the package that's being built by `serverless` contains Mac specific binding. But we need it to be for the environment in which Lambda is running, which is described here - https://docs.aws.amazon.com/lambda/latest/dg/current-supported-versions.html

**What to do**
We need to build the package (especially the binding node_libcurl.node) with an environment similar to AWS Lambda. The best way is to use the Lambda docker image (https://github.com/lambci/docker-lambda). Here is a small `Dockerfile` for the nodejs8.10 environment:
```
# Use docker images for lambda(https://github.com/lambci/docker-lambda)
FROM lambci/lambda:build-nodejs8.10

# Install yarn
RUN curl --silent --location https://dl.yarnpkg.com/rpm/yarn.repo | tee /etc/yum.repos.d/yarn.repo \
    && yum install -y yarn

# Build node-libcurl project
RUN git clone https://github.com/JCMais/node-libcurl.git \
    && cd node-libcurl && yarn install && yarn pregyp rebuild
```
And a shell script to create a container using this Dockerfile, build the package in that container and copy the binding outside -
```
#!/bin/bash

# Remove existing containers and images
docker image rm node-libcurl-amazon-build
docker rm node-libcurl-amazon-build

# Create image from dockerfile
docker build . -t node-libcurl-amazon-build

# Create container and copy node addon
docker create --name node-libcurl-amazon-build node-libcurl-amazon-build
docker cp node-libcurl-amazon-build:/var/task/node-libcurl/lib/binding/node_libcurl.node ./node_libcurl.node

# Remove container and image
docker rm node-libcurl-amazon-build
docker image rm node-libcurl-amazon-build
```
***How to deploy***
If you are using `serverless cli` or deploying using the AWS UI, copy the file `node_libcurl.node`, build by above shell script into `node_modules/node-libcurl/lib/binding` folder (replacing the binding already there). Then you can package the project and upload using the UI or use `serverless deploy`.

If using a deployment script, then add the following `cp` command after `npm install`
```
cp <path to the AWS Lambda node_libcurl.node> ./node_modules/node-libcurl/lib/binding
```

**Comment by tnolet** [comment]
@bairathirahul this is super super awesome. Was looking for exactly this.
